evm-ds/src/precompiles/scilla_call.rs

```diff
@@ -0,0 +1,132 @@
+use ethabi::param_type::ParamType;
+use ethabi::token::Token;
+use hex::ToHex;
+use serde_json::{json, Value};
+
+const BASE_COST: u64 = 15;
+const PER_BYTE_COST: u64 = 3;
+
+pub(crate) fn scilla_call(
+    input: &[u8],
+    gas_limit: Option<u64>,
+    _contex: &Context,
+    backend: &dyn Backend,
+    _is_static: bool,
+) -> Result<(PrecompileOutput, u64), PrecompileFailure> {
+    let gas_needed = match required_gas(input) {
+        Ok(i) => i,
+        Err(err) => return Err(PrecompileFailure::Error { exit_status: err }),
+    };
+
+    if let Some(gas_limit) = gas_limit {
+        if gas_limit < gas_needed {
+            return Err(PrecompileFailure::Error {
+                exit_status: ExitError::OutOfGas,
+            });
+        }
+    }
+
+    let (code_address, passed_transition_name) = get_contract_addr_and_name(input)?;
+
+    let code = backend.code_as_json(code_address);
+    if code.is_empty() {
+        return Err(PrecompileFailure::Error {
+            exit_status: ExitError::Other(Cow::Borrowed("There no code under given address")),
+        });
+    }
+    let Ok(code) = serde_json::from_slice::<serde_json::Value>(&code) else {
+        return Err(PrecompileFailure::Error {
+            exit_status: ExitError::Other(Cow::Borrowed("Unable to parse scilla contract code")),
+        });
+    };
+
+    let transitions = code["contract_info"]["transitions"].to_owned();
+    let Some(transitions) = transitions.as_array() else {
+        return Err(PrecompileFailure::Error {
+            exit_status: ExitError::Other(Cow::Borrowed("Unable to get transitions array from contract json")),
+        });
+    };
+    let mut output_json = build_result_json(input, &passed_transition_name, transitions)?;
+    output_json["_address"] = Value::String(code_address.encode_hex());
+
+    Ok((
+        PrecompileOutput {
+            output_type: PrecompileOutputType::Trap,
+            output: serde_json::to_vec(&output_json).unwrap(),
+        },
+        gas_needed,
+    ))
+}
+
+fn build_result_json(
+    input: &[u8],
+    expected_transition: &str,
+    transitions: &Vec<Value>,
+) -> Result<Value, PrecompileFailure> {
+    let mut solidity_args = vec![ParamType::Address, ParamType::String];
+    let mut scilla_args = vec![];
+
+    for transition in transitions {
+        let name = transition["vname"].as_str().unwrap_or_default();
+        if name.eq(expected_transition) {
+            let Some(args) = transition["params"].as_array() else {
+                return Err(PrecompileFailure::Error {
+                    exit_status: ExitError::Other(Cow::Borrowed("Unable to get transition args from contract json")),
+                });
+            };
+            for arg_obj in args {
+                scilla_args.push((
+                    arg_obj["vname"].as_str().unwrap_or_default(),
+                    arg_obj["type"].as_str().unwrap_or_default(),
+                ));
+            }
+            break;
+        }
+    }
```

Comment:
(contrariwise to the previous comment :-) ) - I think it should probably be an error if you call a non-existent transition, because otherwise there seems to be no way to distinguish between calling an empty transition with no arguments and no return value, which succeeds, and calling a transition which doesn't exist in the target contract?

Definitely worth some comments here to indicate what you think the semantics ought to be..